.github/workflows/integration-test.yml

```diff
@@ -30,14 +35,39 @@ jobs:
       fail-fast: false
 
     steps:
-      - uses: actions/checkout@v4
+      - name: Checkout source
+        if: ${{ github.event_name != 'pull_request_target' }}
+        uses: actions/checkout@v4
+      - name: Fetch user permission
+        if: ${{ github.event_name == 'pull_request_target' }}
+        id: permission
+        uses: actions-cool/check-user-permission@v2
+        with:
+          require: write
+          username: ${{ github.triggering_actor }}
+      - name: Check user permission
+        if: ${{ github.event_name == 'pull_request_target' && steps.permission.outputs.require-result == 'false' }}
+        # If the triggering actor does not have write permission (i.e., this is a
+        # PR from a fork), then we exit, otherwise most of the integration tests will
+        # fail because they require access to secrets.  In this case, a maintainer
+        # will need to make sure the PR looks safe, and if so, manually re-run the
+        # failed jobs.
+        run: |
+          echo "User ${{ github.triggering_actor }} does not have permission to run integration tests."
+          echo "A maintainer must perform a security review and re-run this build, if the code is safe."
+          exit 1
+      - name: Checkout source
+        if: ${{ github.event_name == 'pull_request_target' }}
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ github.event.pull_request.head.sha }}
```

Comment:
```suggestion
      - name: Fetch user permission
        if: ${{ github.event_name == 'pull_request_target' }}
        id: permission
        uses: actions-cool/check-user-permission@v2
        with:
          require: write
          username: ${{ github.triggering_actor }}
      - name: Check user permission
        if: ${{ github.event_name == 'pull_request_target' && steps.permission.outputs.require-result == 'false' }}
        # If the triggering actor does not have write permission (i.e., this is a
        # PR from a fork), then we exit, otherwise most of the integration tests will
        # fail because they require access to secrets.  In this case, a maintainer
        # will need to make sure the PR looks safe, and if so, manually re-run the
        # failed jobs.
        run: |
          echo "User ${{ github.triggering_actor }} does not have permission to run integration tests."
          echo "A maintainer must perform a security review and re-run this build, if the code is safe."
          exit 1
      - name: Checkout source
        uses: actions/checkout@v4
        with:
          ref: ${{ github.event.pull_request.head.sha }}
```

I think we can simplify to one checkout -- if the event_name is pull_request_target, we'll fail before we reach that point due to the permissions checks.

EDIT: What a repulsive diff output. I only removed the first checkout, and removed the conditional from the second checkout.

Reply:
Yes, I think we can simplify this, but we don't want to only one checkout because a "push" event will not have a pull request sha. I'll rework.

Reply:
Since the default for `ref` is `""` (IIRC), can we do `${{ github.event.pull_request.head.sha | '' }}`?